pkg/api/endpoints/util.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpoints
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"hash"
+	"sort"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+// RepackSubsets takes a slice of EndpointSubset objects, expands it to the full
+// representation, and then repacks that into the canonical layout.  This
+// ensures that code which operates on these objects can rely on the common
+// form for things like comparison.  The result is a newly allocated slice.
+func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
+	// First map each unique port definition to the sets of hosts that
+	// offer it.  The sets of hosts must be de-duped, using IP as the key.
+	type ipString string
+	allAddrs := map[ipString]*api.EndpointAddress{}
+	portsToAddrs := map[api.EndpointPort]addressSet{}
+	for i := range subsets {
+		for j := range subsets[i].Ports {
+			epp := &subsets[i].Ports[j]
+			for k := range subsets[i].Addresses {
+				epa := &subsets[i].Addresses[k]
+				ipstr := ipString(epa.IP)
+				// Accumulate the most "complete" address we can.
+				if allAddrs[ipstr] == nil {
+					// Make a copy so we don't write to the
+					// input args of this function.
+					p := &api.EndpointAddress{}
+					*p = *epa
+					allAddrs[ipstr] = p
```

Comment:
What is DeepCopy?  I find no such anything in our tree or in go's libs.
Sadly, this is just something go does not handle well.  No such thing as a
copy-ctor

On Thu, Mar 26, 2015 at 10:05 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/api/endpoints/util.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5939#discussion_r27276206>
> :
>
> > +	type ipString string
> > +	allAddrs := map[ipString]*api.EndpointAddress{}
> > +	portsToAddrs := map[api.EndpointPort]addressSet{}
> > +	for i := range subsets {
> > +		for j := range subsets[i].Ports {
> > +			epp := &subsets[i].Ports[j]
> > +			for k := range subsets[i].Addresses {
> > +				epa := &subsets[i].Addresses[k]
> > +				ipstr := ipString(epa.IP)
> > +				// Accumulate the most "complete" address we can.
> > +				if allAddrs[ipstr] == nil {
> > +					// Make a copy so we don't write to the
> > +					// input args of this function.
> > +					p := &api.EndpointAddress{}
> > +					*p = *epa
> > +					allAddrs[ipstr] = p
>
> Is this really better than DeepCopy?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5939/files#r27276206>
> .
>